### PR TITLE
mrc-4449: Add code generation for parameter types

### PR DESCRIPTION
**Pull request by r-ash**
I think this is big enough as is so will leave the generation of state types to another PR.

After this PR if someone from science wants to add a new input data type it should just be a job of updating the `model_input.csv`, running the `./scripts/generate` and it should be ready to use.

This PR will
* From the `model_input.csv` generate out the parameter types. i.e. `Demography`, `NaturalHistory` etc. How these structs fit into the different model variants is still done by hand but expecting that should change less frequently. I'm doing this by changing the `cpp_name` column to be qualified by the struct it belongs to, at the moment this only supports 1 level deep but we could support more nesting in the future.

This also has a bit of tidying of the generation code.

The files in the `inst/include` dir are starting to get a bit unwieldy, I will look at tidying those up in another PR. Perhaps putting generated code into a separate dir?

### CODE REVIEW

File: inst/cpp_generation/state_types.hpp.in [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+      h_art_adult;
+  real_type births;
+  TensorFixedSize <real_type, Sizes<hDS<ModelVariant>, hAG<ModelVariant>, NS<ModelVariant>>> h_hiv_deaths_no_art;
+  TensorFixedSize <real_type, Sizes<pAG<ModelVariant>, NS<ModelVariant>>> p_infections;
+  TensorFixedSize <real_type, Sizes<hTS<ModelVariant>, hDS<ModelVariant>, hAG<ModelVariant>, NS<ModelVariant>>>
+      h_hiv_deaths_art;
+  TensorFixedSize <real_type, Sizes<hDS<ModelVariant>, hAG<ModelVariant>, NS<ModelVariant>>> h_art_initiation;
+  TensorFixedSize <real_type, Sizes<pAG<ModelVariant>, NS<ModelVariant>>> p_hiv_deaths;
+
+  BaseModelState(const Parameters<ModelVariant, real_type> &pars) {
```
Comment (r-ash) [inline inst/cpp_generation/state_types.hpp.in:32]: I moved this here as part of this PR when I was thinking I would do the output and input generation together. But I will add the generation part of this in another PR

File: inst/cpp_generation/model_input.csv [hunk 2]
```diff
@@ -1,38 +1,38 @@
...
-paed_art_val,hc_art_val,real_type,ModelVariant::run_child_model,,FALSE,1,proj_years,,,
-init_art_dist,hc_art_init_dist,real_type,ModelVariant::run_child_model,,FALSE,2,options.p_idx_hiv_first_adult,proj_years,,
+basepop,Demography.base_pop,real_type,,,FALSE,2,base.pAG,base.NS,,
+Sx,Demography.survival_probability,real_type,,,FALSE,3,base.pAG + 1,base.NS,proj_years,
+netmigr_adj,Demography.net_migration,real_type,,,FALSE,3,base.pAG,base.NS,proj_years,
+asfr,Demography.age_specific_fertility_rate,real_type,,,FALSE,2,options.p_fertility_age_groups,proj_years,,
+births_sex_prop,Demography.births_sex_prop,real_type,,,FALSE,2,base.NS,proj_years,,
+incidinput,Incidence.total_rate,real_type,,,FALSE,1,proj_years,,,
+incrr_age,Incidence.relative_risk_age,real_type,,,FALSE,3,base.pAG - options.p_idx_hiv_first_adult,base.NS,proj_years,
+incrr_sex,Incidence.relative_risk_sex,real_type,,,FALSE,1,proj_years,,,
```
Comment (r-ash) [inline inst/cpp_generation/model_input.csv:9]: @mwalte10 just FYI updated the name of this variable as I think I missed it off when doing big renaming. I think this now matches what we settled on in the README.

Note also the change in format here, now need to specify what struct this data belongs to

**Comment by richfitz** [comment]
https://github.com/mrc-ide/frogger/actions/runs/6585707065/job/17892645869?pr=38 - looks like the leak check action does not error on leaks; there's a crash on one of the builds, unfortunately

**Review by M-Kusumgar — COMMENTED** [review]
Code looks good to me, mostly trying to follow the structure of this project, small suggestion, theres a lot of `paste(..., collapse = "\n")` perhaps extracting that as a helper function `pasteLines` or something like that

**Review by M-Kusumgar — APPROVED** [review]

**Pull request merged**